rpc/connection.go

```diff
@@ -609,6 +609,9 @@ func (c *Connection) doReconnect(ctx context.Context, disconnectStatus Disconnec
 			// A fatal error happened.
 			*reconnectErrPtr = err
 			// short-circuit Retry
+			if err != nil {
+				c.transport.Close()
```

Comment:
Also, this is easy to repro if you just force `OnConnect` to always return an error, and `ShouldRetryOnConnect` to always return false. You should see the process gradually assumulating closed fds.